092-self-healing-using-cruise-control.md

```diff
@@ -0,0 +1,284 @@
+ triggeredSelfHealingTaskId: 4b1ed191-50d7-4aa1-806b-b8c5c0d679fe, triggeredTaskReason:
+ Self healing for DISK_FAILURE: {
+	Disk /var/lib/kafka/data-1/kafka-log2 on broker 2 failed at 2024-12-09T09:49:46Z
+ }}
+```
+
+### Enabling self-healing configuration in the Strimzi operator
+
+Every configuration related to self-healing is currently disabled in the operator, and we will need to enable users altering some of those properties.
+The setting of these properties will be placed behind a feature gate.
+
+### Feature Gate
+
+The new feature gate will be called `UseSelfHealing`.
+It will be introduced in an alpha state and will be disabled by default.
+
+We are using a feature gate for this feature due to various reasons:
+* We want to allow users to try this feature and test it.
+* We would like to know more about the positive/negatives of this feature by testing it in real deployments.
+* Getting feedback from the community would allow us to understand how we can improve this even more.
+
+### Enabling self-healing in the Kafka custom resource
+
+When the feature gate is enabled, the operator will set the `self.healing.enabled` Cruise Control configuration to `true` and allow the users to set other appropriate `self-healing` related [configurations](https://github.com/linkedin/cruise-control/wiki/Configurations#selfhealingnotifier-configurations).
+
+Here is an example of how the configured Kafka custom resource could look:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    config:
+      # configuration to enable self-healing for all anomalies
+      self.healing.enabled: true
+```
+
+Where the user wants to enable self-healing for some specific anomalies, they can use configurations like `self.healing.broker.failure.enabled` for broker failure, `self.healing.goal.violation.enabled` for goal violation etc.
+
+### Strimzi Notifier
+
+Cruise Control provides `AnomalyNotifier` interface which has multiple abstract methods on what to do if certain anomalies are detected.
+`SelfHealingNotifier` is the base class that contains the logic of self-healing and override the methods in by implementing the `AnomalyNotifier` interface. 
+Some of those methods are:`onGoalViolation()`, `onBrokerFailure()`, `onDisk Failure`, `alert()` etc.
+The `SelfHealingNotifier` class can then be further used to implement your own notifier
+The other custom notifier offered by Cruise control are also based upon the `SelfHealingNotifier`.
+
+We will be using our own notifier implementation called `StrimziNotifer`(set via the `anomaly.notifier.class` config).
+Strimzi will configure Cruise Control to always use the `StrimziNotifier` to notify regarding the anomalies and take action on them.
+
+For effective alerts and monitoring, we will be making use of Kubernetes events to signal to users what operations Cruise Control is performing automatically.
```

Comment:
* Why do you think Kubernetes Events are a good way to communicate the anomalies?
* Why should it be forced on the user? 
* Why should it be called `StrimziNotifier`? It seems to be more Strimzi `KubernetesEventsNotifier` as it has nothing really specific to Strimzi?

Reply:
IIRC (during an offline discussion) we went towards the Kube events as a better solution to make users aware (at least) that an anomaly was detected instead of using, for example, a specific custom resource (i.e. `KafkaAnomaly`) or some other Kube built-in resource. How would you make the user aware of detected anomalies? Just logs? AFAIK you should be able to gather Kube events in a Grafana dashboard for monitoring as well.

On the naming side, maybe yes, a more generic `KubernetesEventsNotifier` is better.